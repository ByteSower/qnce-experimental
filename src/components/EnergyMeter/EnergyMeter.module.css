/* EnergyMeter Component Styles */
.energyMeter {
  padding: 1rem;
  background-color: #f1f5f9;
  border-radius: 0.5rem;
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.05);
  border: 1px solid #e2e8f0;
}

.energyLabel {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 0.5rem;
}

.energyText {
  font-size: 0.875rem;
  font-weight: 500;
  color: #374151;
}

.resetButton {
  font-size: 0.75rem;
  color: #64748b;
  text-decoration: underline;
  background: none;
  border: none;
  cursor: pointer;
  border-radius: 0.25rem;
  padding: 0.125rem 0.25rem;
}

.resetButton:hover {
  color: #374151;
}

.resetButton:focus {
  outline: none;
  box-shadow: 0 0 0 2px #cbd5e1;
}

.progressContainer {
  position: relative;
  background-color: #e2e8f0;
  border-radius: 9999px;
  overflow: hidden;
}

.progressFill {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  border-radius: 9999px;
  transition: all 0.5s ease-out;
}

.progressShine {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  background: linear-gradient(90deg, transparent, rgba(255, 255, 255, 0.2), transparent);
  transition: all 0.5s;
}

.harvestContainer {
  margin-top: 0.75rem;
  display: flex;
  justify-content: center;
}

.harvestButton {
  padding: 0.5rem 1rem;
  border-radius: 0.375rem;
  font-weight: 500;
  font-size: 0.875rem;
  transition: all 0.2s;
  border: none;
  cursor: pointer;
}

.harvestButton:focus {
  outline: none;
  box-shadow: 0 0 0 2px rgba(34, 197, 94, 0.5);
}

.harvestButtonEnabled {
  background-color: #22c55e;
  color: white;
}

.harvestButtonEnabled:hover {
  background-color: #16a34a;
  transform: scale(1.05);
}

.harvestButtonEnabled:active {
  transform: scale(0.95);
}

.harvestButtonDisabled {
  background-color: #d1d5db;
  color: #6b7280;
  cursor: not-allowed;
}

.statusIndicator {
  margin-top: 0.5rem;
  text-align: center;
}

.statusText {
  font-size: 0.75rem;
}

.statusTextMaxed {
  color: #ea580c;
  font-weight: 500;
}

.statusTextNormal {
  color: #64748b;
}

/* Energy level color classes */
.energyFillLow {
  background: linear-gradient(90deg, #4ade80, #22c55e);
}

.energyFillMedium {
  background: linear-gradient(90deg, #facc15, #eab308);
}

.energyFillHigh {
  background: linear-gradient(90deg, #fb923c, #f97316);
}
