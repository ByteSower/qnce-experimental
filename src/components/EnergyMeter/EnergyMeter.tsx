/**
 * PLAN: EnergyMeter component
 * - Import and use useEnergy hook.
 * - Render a progress bar showing energy vs cap.
 * - Display numeric label "GP: {energy} / {cap}".
 * - Harvest button calls addEnergy(1), disabled at isMaxed.
 * - Reset link calls resetEnergy().
 * - Animate bar fill on energy change.
 * - Accept optional props: width, height, labelFormat.
 * - Add ARIA roles for progressbar and button.
 */
// Copilot: Implement EnergyMeter component in React + TypeScript per the above plan.
// Use CSS modules for styling to avoid inline styles.
// Export: EnergyMeter(props?: { width?: number; height?: number; labelFormat?: (energy:number,cap:number)=>string; })

import React, { useEffect, useRef } from 'react';
import { useEnergy } from '../../hooks/useEnergy';
import type { UseEnergyReturn } from '../../hooks/useEnergy';
import styles from './EnergyMeter.module.css';

interface EnergyMeterProps {
  width?: number;
  height?: number;
  labelFormat?: (energy: number, cap: number) => string;
  config?: {
    start?: number;
    rate?: number;
    intervalMs?: number;
    cap?: number;
  };
  // Optional external energy state
  energyState?: UseEnergyReturn;
}

export const EnergyMeter: React.FC<EnergyMeterProps> = ({
  width = 300,
  height = 24,
  labelFormat = (energy, cap) => `GP: ${energy} / ${cap}`,
  config = {},
  energyState
}) => {
  // Use external energy state if provided, otherwise create internal state
  const internalEnergyState = useEnergy(config);
  const { energy, addEnergy, resetEnergy, isMaxed } = energyState || internalEnergyState;
  const cap = config.cap || 100;
  const containerRef = useRef<HTMLDivElement>(null);
  const fillRef = useRef<HTMLDivElement>(null);
  const shineRef = useRef<HTMLDivElement>(null);
  
  // Calculate fill percentage
  const fillPercentage = Math.min((energy / cap) * 100, 100);
  
  // Calculate color class based on energy level
  const getEnergyColorClass = (): string => {
    const percentage = fillPercentage;
    if (percentage < 30) {
      return styles.energyFillLow;
    } else if (percentage < 70) {
      return styles.energyFillMedium;
    } else {
      return styles.energyFillHigh;
    }
  };

  // Update dynamic styles via refs
  useEffect(() => {
    if (containerRef.current) {
      containerRef.current.style.width = `${width}px`;
      containerRef.current.style.height = `${height}px`;
    }
    if (fillRef.current) {
      fillRef.current.style.width = `${fillPercentage}%`;
    }
    if (shineRef.current) {
      shineRef.current.style.width = `${fillPercentage}%`;
    }
  }, [width, height, fillPercentage]);

  const handleHarvest = () => {
    if (!isMaxed) {
      addEnergy(1);
    }
  };

  const handleReset = () => {
    resetEnergy();
  };

  return (
    <div className={styles.energyMeter}>
      {/* Energy Label */}
      <div className={styles.energyLabel}>
        <span className={styles.energyText}>
          {labelFormat(energy, cap)}
        </span>
        <button
          onClick={handleReset}
          className={styles.resetButton}
          aria-label="Reset energy to starting value"
        >
          Reset
        </button>
      </div>

      {/* Progress Bar Container */}
      <div
        ref={containerRef}
        className={styles.progressContainer}
        role="progressbar"
        aria-valuenow={Math.round(energy)}
        aria-valuemin={0}
        aria-valuemax={cap}
        aria-label={`Energy level: ${energy} out of ${cap} growth points`}
      >
        {/* Progress Bar Fill */}
        <div
          ref={fillRef}
          className={`${styles.progressFill} ${getEnergyColorClass()}`}
        />
        
        {/* Shine effect overlay */}
        <div
          ref={shineRef}
          className={styles.progressShine}
        />
      </div>

      {/* Harvest Button */}
      <div className={styles.harvestContainer}>
        <button
          onClick={handleHarvest}
          disabled={isMaxed}
          className={`
            ${styles.harvestButton}
            ${isMaxed ? styles.harvestButtonDisabled : styles.harvestButtonEnabled}
          `}
          aria-label={isMaxed ? 'Cannot harvest - energy is at maximum' : 'Harvest light to gain 1 growth point'}
        >
          {isMaxed ? 'Energy Full' : 'ðŸŒ± Harvest Light'}
        </button>
      </div>

      {/* Status Indicator */}
      <div className={styles.statusIndicator}>
        <span className={`${styles.statusText} ${isMaxed ? styles.statusTextMaxed : styles.statusTextNormal}`}>
          {isMaxed ? 'âœ¨ Maximum Energy Reached!' : 'Collecting solar energy...'}
        </span>
      </div>
    </div>
  );
};
