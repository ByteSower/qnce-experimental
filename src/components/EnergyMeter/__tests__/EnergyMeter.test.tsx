import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { EnergyMeter } from '../EnergyMeter';

describe('EnergyMeter Component', () => {
  test('renders without crashing', () => {
    render(<EnergyMeter />);
    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });

  test('displays progress bar and buttons', () => {
    render(<EnergyMeter />);
    
    expect(screen.getByRole('progressbar')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /harvest/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /reset/i })).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { container } = render(<EnergyMeter />);
    expect(container.firstChild).toMatchSnapshot();
  });
});