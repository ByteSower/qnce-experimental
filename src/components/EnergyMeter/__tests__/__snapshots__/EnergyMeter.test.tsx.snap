// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`EnergyMeter Component matches snapshot 1`] = `
<div>
  <div>
    <span>
      GP: 0 / 100
    </span>
    <button
      aria-label="Reset energy to starting value"
    >
      Reset
    </button>
  </div>
  <div
    aria-label="Energy level: 0 out of 100 growth points"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="0"
    role="progressbar"
    style="width: 300px; height: 24px;"
  >
    <div
      class="undefined undefined"
      style="width: 0%;"
    />
    <div
      style="width: 0%;"
    />
  </div>
  <div>
    <button
      aria-label="Harvest light to gain 1 growth point"
      class="
            undefined
            undefined
          "
    >
      ðŸŒ± Harvest Light
    </button>
  </div>
  <div>
    <span
      class="undefined undefined"
    >
      Collecting solar energy...
    </span>
  </div>
</div>
`;
